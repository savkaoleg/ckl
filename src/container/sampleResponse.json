{
  "results": [
    {
      "gender": "female",
      "name": { "title": "mrs", "first": "christina", "last": "bailey" },
      "location": {
        "street": "5144 grove road",
        "city": "stoke-on-trent",
        "state": "northumberland",
        "postcode": "B1 7BJ",
        "coordinates": { "latitude": "50.2501", "longitude": "36.0567" },
        "timezone": {
          "offset": "-11:00",
          "description": "Midway Island, Samoa"
        }
      },
      "email": "christina.bailey@example.com",
      "login": {
        "uuid": "309084db-66a9-43db-b7a1-dd1f5535edd9",
        "username": "beautifulswan866",
        "password": "niceguy",
        "salt": "zNMpZU0C",
        "md5": "8d0765fa836e798f3986d72d7b8f52f5",
        "sha1": "a307b7d11277826a00ccce6a0abfb6d728d94059",
        "sha256": "18f4e50b0e5d99ed593e00cdd75778a472cbd2ecc44f0b8345487a46fb0d25ec"
      },
      "dob": { "date": "1951-02-15T01:04:16Z", "age": 67 },
      "registered": { "date": "2017-09-14T20:35:38Z", "age": 1 },
      "phone": "0101 298 6494",
      "cell": "0782-489-800",
      "id": { "name": "NINO", "value": "KY 39 30 87 G" },
      "picture": {
        "large": "https://randomuser.me/api/portraits/women/30.jpg",
        "medium": "https://randomuser.me/api/portraits/med/women/30.jpg",
        "thumbnail": "https://randomuser.me/api/portraits/thumb/women/30.jpg"
      },
      "nat": "GB"
    },
    {
      "gender": "male",
      "name": { "title": "mr", "first": "gerald", "last": "burns" },
      "location": {
        "street": "855 main street",
        "city": "edinburgh",
        "state": "tayside",
        "postcode": "IU1P 4PJ",
        "coordinates": { "latitude": "63.2009", "longitude": "108.5937" },
        "timezone": {
          "offset": "-6:00",
          "description": "Central Time (US & Canada), Mexico City"
        }
      },
      "email": "gerald.burns@example.com",
      "login": {
        "uuid": "760475b7-4e97-4be7-8399-03eda7f64fdf",
        "username": "whiteswan950",
        "password": "smut",
        "salt": "KbAWNxOX",
        "md5": "cc2067cbd5897268ca655fe01c31caf4",
        "sha1": "f84d1ad61bc1c0fe9bb296a6ce63e997edee80d8",
        "sha256": "713f36bd1b1b6fa79f19a3c3402f147072de3ed6985b137d131ccc51de34122c"
      },
      "dob": { "date": "1987-06-29T00:25:14Z", "age": 31 },
      "registered": { "date": "2016-03-18T06:06:46Z", "age": 2 },
      "phone": "023 5157 4687",
      "cell": "0754-130-528",
      "id": { "name": "NINO", "value": "JB 90 90 81 A" },
      "picture": {
        "large": "https://randomuser.me/api/portraits/men/83.jpg",
        "medium": "https://randomuser.me/api/portraits/med/men/83.jpg",
        "thumbnail": "https://randomuser.me/api/portraits/thumb/men/83.jpg"
      },
      "nat": "GB"
    }
  ],
  "info": {
    "seed": "3bca2bc7d9d7409b",
    "results": 5,
    "page": 1,
    "version": "1.2"
  }
}
