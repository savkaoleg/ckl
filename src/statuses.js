// @flow
const arr = {
  applied: 'applied',
  interviewing: 'interviewing',
  hired: 'hired'
}

export default arr
