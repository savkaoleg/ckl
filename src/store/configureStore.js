import { createStore, applyMiddleware } from 'redux'
import rootReducer from './rootReducer'
import thunkMiddleware from 'redux-thunk'
import { createLogger } from 'redux-logger'
import { composeWithDevTools } from 'redux-devtools-extension'

let middleWares = [thunkMiddleware]

const loggerMiddleware = createLogger({
  collapsed: () => {
    return true
  }
})

const envArr = ['production', 'test']

if (!envArr.includes(process.env.NODE_ENV)) {
  middleWares = [...middleWares, loggerMiddleware]
}

const createStoreWithMiddleware = composeWithDevTools(
  applyMiddleware(...middleWares)
)(createStore)

export default function configureStore(initialState) {
  return createStoreWithMiddleware(rootReducer, initialState)
}
