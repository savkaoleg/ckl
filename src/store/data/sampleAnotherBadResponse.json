{
  "error": "Please, use another version of our api",
  "results": []
}
