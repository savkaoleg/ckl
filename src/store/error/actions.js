// @flow
import { ERROR_CHANGE } from './types'

export function setError(errorMsg) {
  return {
    type: ERROR_CHANGE,
    errorMsg
  }
}
