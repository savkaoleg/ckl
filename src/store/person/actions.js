// @flow
import { ADD_PERSON, CHANGE_PERSON } from './types'

export function addPerson(data: Object) {
  return {
    type: ADD_PERSON,
    data
  }
}

export function changePerson(uuid: string, status: string) {
  return {
    type: CHANGE_PERSON,
    uuid,
    status
  }
}
