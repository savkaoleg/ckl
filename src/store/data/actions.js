// @flow
import axios from 'axios'
import { SET_DATA } from './types'
import { setError } from '../error/actions'
import { setLoaded } from '../loaded/actions'
import statuses from '../../status.json'

export const url = 'https://randomuser.me/api/?nat=gb&results=5'

export class Api {
  constructor(options = {}) {
    this.client = options.client || axios.create()
    this.url = options.url || url
  }

  loadData() {
    return this.client(this.url).then(({ data }) => data)
  }
}

export function setData(payload) {
  return {
    type: SET_DATA,
    payload
  }
}

export function loadData(apiProvider) {
  return async dispatch => {
    try {
      const api = new Api(apiProvider)
      const data = await api.loadData()

      if (data.results) {
        if (data.results.length) {
          const personsForRedux = data.results.map(item => {
            return {
              status: statuses.applied,
              ...item
            }
          })
          dispatch(setData(personsForRedux))
        } else {
          dispatch(setError('Something wrong'))
        }
      } else {
        dispatch(setError('Something wrong'))
      }

      dispatch(setLoaded(true))
    } catch (e) {
      console.log('e', e)
      dispatch(setError(e.message))
      dispatch(setLoaded(true))
    }
  }
}
