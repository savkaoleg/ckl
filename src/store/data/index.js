// @flow
import { SET_DATA } from './types'
import { ADD_PERSON, CHANGE_PERSON } from '../person/types'
import person from '../person/index'

export default function data(state: Array<PersonType> = [], action) {
  const { type, payload } = action
  switch (type) {
    case ADD_PERSON:
      return [...state, person({}, action)]
    case CHANGE_PERSON:
      return state.map(personItem => person(personItem, action))
    case SET_DATA:
      return payload
    default:
      return state
  }
}
