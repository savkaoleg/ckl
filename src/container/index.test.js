import React from 'react'
import PropTypes from 'prop-types'
import axios from 'axios'
import MockAdapter from 'axios-mock-adapter'
import expect from 'expect'
import { mount } from 'enzyme'
import Container from './index'

import { Provider } from 'react-redux'
import configureStore from '../store/configureStore'
import { loadData, url } from '../store/data/actions'
import statuses from '../status.json'
import sampleResponse from './sampleResponse.json'

const El = ({ store }) => (
  <Provider store={store}>
    <Container />
  </Provider>
)

El.propTypes = {
  store: PropTypes.object
}

describe('Container', () => {
  it('Renders correctly', () => {
    const store = configureStore()
    const MountedContainer = mount(<El store={store} />)
    expect(MountedContainer).toMatchSnapshot()
  })

  it('Wait and check if all is good', async () => {
    const store = configureStore()
    const MountedContainer = mount(<El store={store} />)
    await store.dispatch(loadData()) //simulate bootstrap store
    MountedContainer.update()

    const persons = MountedContainer.find('.personTitle')
    expect(persons.length).toBeGreaterThan(1)
  })

  it('Wait and check if all is bad', async () => {
    const client = axios.create()
    const fakeUrl = 'https://radndomuser.me/api/?nat=gb&results=5'
    const store = configureStore()
    const MountedContainer = mount(<El store={store} />)

    await store.dispatch(loadData({ client, url: fakeUrl })) //simulate bootstrap store
    MountedContainer.update()

    expect(store.getState().error).toEqual('Network Error')
  })

  it('Simuatr change person status from left to right', async () => {
    const client = axios.create()
    const mock = new MockAdapter(client)

    mock.onGet(url).reply(() => {
      return [200, sampleResponse]
    })

    const store = configureStore()
    const MountedContainer = mount(<El store={store} />)
    await store.dispatch(loadData({ client })) //simulate bootstrap store
    MountedContainer.update()
    const appliedColum = MountedContainer.find('.' + statuses.applied)
    const persons = appliedColum.find('.card')
    const samplePerson = persons.first()
    const sampleBtn = samplePerson.find('Button')
    sampleBtn.simulate('click')
    MountedContainer.update()
    expect(store.getState().data[0].status).toEqual(statuses.interviewing)
  })

  it('Simuatr change person status from right to left', async () => {
    const client = axios.create()
    const mock = new MockAdapter(client)

    mock.onGet(url).reply(() => {
      return [200, sampleResponse]
    })

    const store = configureStore()
    const MountedContainer = mount(<El store={store} />)
    await store.dispatch(loadData({ client })) //simulate bootstrap store
    MountedContainer.update()
    const appliedColum = MountedContainer.find('.' + statuses.applied)
    const persons = appliedColum.find('.card')
    const samplePerson = persons.first()
    const sampleBtn = samplePerson.find('button')
    sampleBtn.simulate('click')
    MountedContainer.update()

    const interviewingColum = MountedContainer.find('.' + statuses.interviewing)
    const person = interviewingColum.find('.card').first()
    const moveLeftBtn = person.find('Button').find({ text: '<<' })
    moveLeftBtn.simulate('click')
    MountedContainer.update()

    expect(store.getState().data[0].status).toEqual(statuses.applied)
  })
})
