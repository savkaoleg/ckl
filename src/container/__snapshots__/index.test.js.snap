// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container Renders correctly 1`] = `
<El
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <Connect(Component)>
      <Component
        changeStatus={[Function]}
        data={Array []}
        error=""
        loaded={false}
      >
        <Loader
          text="Loading ..."
        >
          <div
            className="showbox"
          >
            <div
              className="loader"
            >
              <svg
                className="circular"
                viewBox="25 25 50 50"
              >
                <circle
                  className="path"
                  cx="50"
                  cy="50"
                  fill="none"
                  r="20"
                  strokeMiterlimit="10"
                  strokeWidth="2"
                />
              </svg>
            </div>
            <h3
              className="loader-title"
            >
              Loading ...
            </h3>
          </div>
        </Loader>
      </Component>
    </Connect(Component)>
  </Provider>
</El>
`;
